---
title: "Ambiguous Colocalization from Trait-Specific Effects"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Ambiguous Colocalization from Trait-Specific Effects}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```


This vignette demonstrates an example of ambiguous colocalization from trait-specific effects using the `colocboost`. 
Specifically, we will use the `Ambiguous_Colocalization`, which is an example result from GTEx relsease v8 and UK Biobank summary statistics
(see more details of data license in Acknowledgement section).

```{r setup}
library(colocboost)
# Run colocboost with diagnostic details
data(Ambiguous_Colocalization)
names(Ambiguous_Colocalization)
```

# 1. The `Ambiguous_Colocalization` Dataset

The `Ambiguous_Colocalization` dataset contains results from a colocboost analysis of a real genomic region showing ambiguous trait-specific effects between eQTL 
(expression quantitative trait loci) and GWAS (genome-wide association study) signals. 
Ambiguous colocalization occurs when there appears to be shared causal variants between traits, 
but the evidence is complicated by the presence of trait-specific effects. This dataset is structured as a list with two main components:

1. `ColocBoost_Results`: Contains the output from running the ColocBoost algorithm.

2. `SuSiE_Results`: Contains fine-mapping results from the SuSiE algorithm for both eQTL and GWAS data separately.



# 2. ColocBoost results

In this example, there are two trait-specifc effects for the eQTL and GWAS signals, respectively. But two uCoS have overlapping variants, 
which indicates that the two uCoS are not independent. ColocBoost identifies two uCoS:

- `ucos1:y1`: eQTL trait-specific effect has 6 variants.
- `ucos2:y2`: GWAS trait-specific effect has 22 variants.
- There are 3 variants that are overlapping between the two uCoS.

```{r colocboost-results}
# Trait-specific effects for both eQTL and GWAS
Ambiguous_Colocalization$ColocBoost_Results$ucos_details$ucos$ucos_index

# Intersection of eQTL and GWAS variants
Reduce(intersect, Ambiguous_Colocalization$ColocBoost_Results$ucos_details$ucos$ucos_index)
```

After checking the correlation of variants between the two uCoS, we can see the high correlation between the two uCoS.

- The minimum absolute correlation between the two uCoS is 0.636 (off-diagonal `purity$min_abs_corr`).
- The median absolute correlation between the two uCoS is 0.837 (off-diagonal `purity$median_abs_corr`).
- The maximum absolute correlation between the two uCoS is 1 (off-diagonal `purity$max_abs_corr`), indicating overlapping variants exists.

```{r colocboost-purity}
# With-in and between purity
Ambiguous_Colocalization$ColocBoost_Results$ucos_details$ucos_purity
```

Based on the results, we can see that the two uCoS are not independent, but they are not fully overlapping.

```{r plot-ambiguous}
n_variables <- Ambiguous_Colocalization$ColocBoost_Results$data_info$n_variables
colocboost_plot(
  Ambiguous_Colocalization$ColocBoost_Results, 
  plot_cols = 1,
  grange = c(2000:n_variables),
  plot_ucos = TRUE,
  show_cos_to_uncoloc = TRUE
)
```



# 2. Fine-mapping results from SuSiE

In this example, we also have fine-mapping results from SuSiE for both eQTL and GWAS data separately.

- For eQTL, SuSiE shows 40 variants in 95% credible set (CS).
- For GWAS, SuSiE shows 57 variants in 95% credible set (CS).
- There are 24 variants are overlapping between the two credible sets.

```{r susie-results}
susie_eQTL <- Ambiguous_Colocalization$SuSiE_Results$eQTL
susie_GWAS <- Ambiguous_Colocalization$SuSiE_Results$GWAS

# Fine-mapped eQTL
susie_eQTL$sets$cs$L1

# Fine-mapped GWAS variants
susie_GWAS$sets$cs$L1

# Intersection of fine-mapped eQTL and GWAS variants
intersect(susie_eQTL$sets$cs$L1, susie_GWAS$sets$cs$L1)
```

To visulize the fine-mapping results,

```{r plot-susie}
susieR::susie_plot(susie_eQTL, y = "PIP")
susieR::susie_plot(susie_GWAS, y = "PIP")
```


# 3. Get the ambiguous colocalization results and summary

ColocBoost provides a function to get the ambiguous colocalization results and summary from trait-specifc effects, by considering the correlation of variants between the two uCoS.

## 3.1. Get the ambiguous colocalization results

The `get_ambiguous_colocalization` function will return the ambiguous results in `ambigous_ucos` object, if the following conditions are met:

- The two uCoS should have at least one overlapping variant.
- The minimum absolute correlation between the two uCoS is greater than `min_abs_corr_between_ucos` (default is 0.5).
- The median absolute correlation between the two uCoS is greater than `median_abs_corr_between_ucos` (default is 0.8).


```{r ambiguous-results}
colocboost_results <- Ambiguous_Colocalization$ColocBoost_Results
res <- get_ambiguous_colocalization(
  colocboost_results, 
  min_abs_corr_between_ucos = 0.5, 
  median_abs_corr_between_ucos = 0.8
)
names(res)
names(res$ambiguous_cos)
names(res$ambiguous_cos[[1]])
```

**Explanation of results** For each ambiguous colocalization, the following information is provided:

- `ambigous_cos`: Contains variants indices and names of the original trait-specific uCoS used to constuct this ambiguouse colocalization.
- `ambigouse_cos_overlap`: Contains the overlapping variants information across the uCoS  used to constuct this ambiguouse colocalization.
- `ambigouse_cos_union`: Contains the union of variants information across the uCoS used to constuct this ambiguouse colocalization.
- `ambigouse_cos_outcomes`: Contains the outcomes indicies and names for uCoS used to constuct this ambiguouse colocalization.
- `ambigous_cos_weight`: Contains the trait-specific weights of the uCoS used to constuct this ambiguouse colocalization.
- `ambigous_cos_puriry`: Contains the purity of across uCoS used to constuct this ambiguouse colocalization.
- `recalibrated_cos_vcp`: Contains the recalibrated integrative weight to analogous to variant colocalization probability (VCP) from the ambiguous colocalization results.
- `recalibrated_cos`: Contains the recalibrated 95% colocalization confidence set (CoS) from the ambiguous colocalization results.


## 3.2. Get the summary of ambiguous colocalization results

To get the summary of ambiguous colocalization results, we can use the `get_colocboost_summary` function.

- `summary_level = 1` (default): get the summary table for only the colocalization results, same as `cos_summary` in ColocBoost output.
- `summary_level = 2`: get the summary table for both colocalization and trait-specific effects if exists.
- `summary_level = 3`: get the summary table for colocalization, trait-specific effects and ambiguous colocalization results if exists.

```{r ambiguous-summary}
# Get the full summary results from colocboost
full_summary <- get_colocboost_summary(colocboost_results, summary_level = 3)
names(full_summary)

# Get the summary of ambiguous colocalization results
summary_ambiguous <- full_summary$ambiguous_cos_summary
colnames(summary_ambiguous)
```

- `recalibrated_*`: giving the recalibrated weigths and recalibrated 95% colocalization confidece sets (CoS) from the trait-specific effects.

See details of function usage in the [Functions](https://statfungen.github.io/colocboost/reference/index.html).



# 4. Take home message

In this vignette, we have demonstrated how the post-processing of ColocBoost results to handle ambiguous colocalization scenarios 
where trait-specific effects share overlapping variants with high correlation. 

- Through empirical studies, we have found that the minimum absolute correlation cutoff of 0.5 and median absolute correlation cutoff of 0.8 produce reasonable results 
when merging two uCoS. We recommend users not to lower these thresholds further without strong justification. 
- While we provide recalibrated weights as a suggested approach for interpreting ambiguous results, 
users can still choose between recalibrated weights and trait-specific weights based on their research context. 
- This ambiguity typically arises when some trait-specific boosting learners are updating very similar sets of variants, 
leading to substantial correlation between what initially appear to be separate signals. 

By properly identifying and quantifying this ambiguity, users can gain a more nuanced understanding of the underlying genetic architecture of the traits they are studying.


# Acknowledgements

- The eQTL data used for the analyses described in this example results were obtained from GTEx release v8 from [GTEx Portal](https://gtexportal.org/home/downloads/adult-gtex/qtl).
- The GWAS summary statistics used for the analyses described in this example results were obtained from UK Biobank (UKBB) 
from [UK Biobank](https://www.ukbiobank.ac.uk/) and downloaded from [Nealelab](https://www.nealelab.is/uk-biobank).