---
title: "ColocBoost Tutorial (Advanced Usage)"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{ColocBoost Tutortial (Advanced Usage)}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(colocboost)
```

## Leveraging a single genotype or LD matrix across multiple phenotypes or summary statistics

In large-scale genetic studies, particularly those involving colocalization analysis of gene expression across different tissues or cell types (such as GTEx or ROSMAP), it is common to encounter scenarios where a single genotype matrix applies to multiple phenotypes. Similarly, for summary statistics analyses involving multiple complex diseases (like those from UK Biobank), researchers often use a single LD matrix across various datasets. Traditional methods/usage that require duplicating these matrices for each phenotype or summary statistics not only increase computational load but also significantly waste memory resources. 

To address these inefficiencies, ColocBoost offers a streamlined option that supports using a single genotype matrix alongside a matrix of phenotypes or a single LD matrix with a list of summary statistics dataframes. This approach leverages the consistency across genetic structures to reduce redundancy and enhance computational efficiency.


**Example of using individual-level data**

Consider a scenario where all phenotypes share the same genotype matrix, as demonstrated with the `Ind_5traits` dataset. Instead of loading multiple copies of the genotype matrix, we use the first genotype matrix for all phenotypes and covert list of phenotype to be a single matrix.

```{r oneGeno}
data("Ind_5traits")
X <- Ind_5traits$X[[1]]
Y <- do.call(cbind, Ind_5traits$Y)
res <- colocboost(X = X, Y = Y)
res$summary_table
```

**Example of using summary statistics**

For summary statistics, similar optimization is applied using `Sumstat_5traits` dataset where all statistical summaries share the same LD matrix.

```{r oneLD}
data("Sumstat_5traits")
sumstat <- Sumstat_5traits$sumstat
LD <- Sumstat_5traits$LD
res <- colocboost(sumstat = sumstat, LD = LD)
res$summary_table
```

**Example of combining individual-level data and summary statistics**

Consider a scenario where several individual-level phenotypes share a common genotype matrix, while multiple summary statistics share the same LD matrix. Our goal is to identify the colocalization across this mixture of individual-level phenotypes and multiple summary statistics.


```{r oneMixture}
data("Ind_5traits")
X <- Ind_5traits$X[[1]]
Y <- do.call(cbind, Ind_5traits$Y[1:3])
data("Sumstat_5traits")
sumstat <- Sumstat_5traits$sumstat[4:5]
LD <- Sumstat_5traits$LD
res <- colocboost(X = X, Y = Y, sumstat = sumstat, LD = LD)
res$summary_table
```


## Few genotype matrices or few LD matrices than phenotypes and summary statistics (dict_YX and dict_sumstat)

## Different phenotypes has different SNPs.



